### PR TITLE
Subgraph Parallelism support for Transformer models

### PR DESCRIPTION
**Pull request by aj-prime**
This PR exposes subgraph parallelism as a new strategy to parallelize the DNN training by processing different branches on separate GPUs.

It implements Data Sub-Graph Parallelism (D&SP) and D&SP-cSub designs. 

Initial support is available to extend sub-graph parallelism for common layers in Transformer models to train out-of-core transformer models. 

An example is available in "applications/nlp/transformer/subgraph" directory. 

Note: It does not support multiple trainers and needs "translatebetweengrid" functions in Hydrogen.

### CODE REVIEW

File: include/lbann/layers/transform/dist_SOB.hpp [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+};
+
+
+
+LBANN_DEFINE_LAYER_BUILDER(dist_SOB);
+
+#ifndef LBANN_DIST_SOB_LAYER_INSTANTIATE
+#define PROTO_DEVICE(T, Device) \
+  extern template class dist_SOB_layer<T, data_layout::DATA_PARALLEL, Device>; \
+  extern template class dist_SOB_layer<T, data_layout::MODEL_PARALLEL, Device>
```
Comment (timmoon10) [inline include/lbann/layers/transform/dist_SOB.hpp:214]: Model-parallel case is not supported.
```suggestion
  extern template class dist_SOB_layer<T, data_layout::DATA_PARALLEL, Device>;
```

File: src/models/model.cpp [hunk 2]
```diff
@@ -1078,13 +1962,16 @@ void model::update_weights() {
...
+  int counter = 0;
   for (auto rit = m_weights.rbegin(); rit != m_weights.rend(); ++rit) {
     auto& w = **rit;
     auto&& opt = w.get_optimizer();
+
     if (opt != nullptr) {
       do_weight_optimize_begin_cbs(&w);
       opt->step();
       do_weight_optimize_end_cbs(&w);
+      counter++;
```
Comment (timmoon10) [inline src/models/model.cpp:1974]: `counter` doesn't do anything:
```suggestion
  for (auto rit = m_weights.rbegin(); rit != m_weights.rend(); ++rit) {
    auto& w = **rit;
    auto&& opt = w.get_optimizer();
    if (opt != nullptr) {
      do_weight_optimize_begin_cbs(&w);
      opt->step();
      do_weight_optimize_end_cbs(&w);
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]
I've raised a few correctness issues that should be addressed before merging. It looks like handles the use-cases from the paper well, although it's all very kludgy. It's not __too__ terrible as long as this is an expert feature, although we should clean things up at some point in the future.

File: include/lbann/layers/layer.hpp [hunk 3]
```diff
@@ -636,6 +715,12 @@ class Layer {
    */
   std::string m_name;
 
+  int enable_vector_copy = 0;
```
[the remark this answers is not in the document]
Comment (aj-prime) [reply include/lbann/layers/layer.hpp:766]: This variable can have multiple values. Ranging from 0 to 2.

**Review by timmoon10 — APPROVED** [review]
Pending no new Bamboo failures, this is ready to merge.

**Comment by timmoon10** [comment]
@mrwyattii Do we see any build errors on Corona?

**Comment by timmoon10** [comment]
[No new Bamboo errors](https://lc.llnl.gov/bamboo/browse/LBANN-TIM377-8).

**Pull request merged**